x/superfluid/keeper/migrate_test.go

```diff
@@ -616,6 +629,73 @@ func (s *KeeperTestSuite) TestMigrateNonSuperfluidLockBalancerToConcentrated() {
 	}
 }
 
+func (s *KeeperTestSuite) TestMigrateUnlockedPositionFromBalancerToConcentrated() {
+	defaultJoinTime := s.Ctx.BlockTime()
+	type sendTest struct {
+		unlocking                bool
+		percentOfSharesToMigrate sdk.Dec
+		tokenOutMins             sdk.Coins
+		expectedError            error
+	}
+	testCases := map[string]sendTest{
+		"no lock (full shares)": {
+			percentOfSharesToMigrate: sdk.MustNewDecFromStr("1"),
+		},
+		"no lock (partial shares)": {
+			percentOfSharesToMigrate: sdk.MustNewDecFromStr("0.9"),
+		},
+		"error: no lock (full shares), token out mins is more than exit coins": {
+			percentOfSharesToMigrate: sdk.MustNewDecFromStr("1"),
+			tokenOutMins:             sdk.NewCoins(sdk.NewCoin("foo", sdk.NewInt(10000))),
+			expectedError:            gammtypes.ErrLimitMinAmount,
+		},
+	}
+
+	for name, tc := range testCases {
+		s.Run(name, func() {
+			s.SetupTest()
+			s.Ctx = s.Ctx.WithBlockTime(defaultJoinTime)
+			ctx := s.Ctx
+			superfluidKeeper := s.App.SuperfluidKeeper
+			gammKeeper := s.App.GAMMKeeper
+
+			// We bundle all migration setup into a single function to avoid repeating the same code for each test case.
+			joinPoolAmt, _, balancerLock, _, poolJoinAcc, balancerPooId, clPoolId, balancerPoolShareOut, _ := s.SetupMigrationTest(ctx, false, false, false, true, tc.percentOfSharesToMigrate)
+			s.Require().Equal(uint64(0), balancerLock.GetID())
+
+			// Depending on the test case, we attempt to migrate a subset of the balancer LP tokens we originally created.
+			coinsToMigrate := balancerPoolShareOut
+			coinsToMigrate.Amount = coinsToMigrate.Amount.ToDec().Mul(tc.percentOfSharesToMigrate).RoundInt()
+
+			// RouteMigration is called via the migration message router and is always run prior to the migration itself
+			synthLockBeforeMigration, migrationType, err := superfluidKeeper.RouteMigration(ctx, poolJoinAcc, 0, coinsToMigrate)
+			s.Require().NoError(err)
+			s.Require().Equal((lockuptypes.SyntheticLock{}), synthLockBeforeMigration)
+			s.Require().Equal(migrationType, keeper.Unlocked)
+
+			// System under test.
+			positionId, amount0, amount1, _, _, poolIdLeaving, poolIdEntering, err := gammKeeper.MigrateUnlockedPositionFromBalancerToConcentrated(ctx, poolJoinAcc, coinsToMigrate, tc.tokenOutMins)
```

Comment:
Should we be checking the liquidity and jointime return values as well?

Reply:
Yeah good call, added this to the other tests too! https://github.com/osmosis-labs/osmosis/pull/5360/commits/c679ac2d85469476775aff1ea815dc3c16fd0c10